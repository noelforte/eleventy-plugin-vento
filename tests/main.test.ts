import { describe, test } from 'vitest';
import { EleventyTest } from './_eleventy-test-instance.js';

const testInstance = new EleventyTest('./tests/stubs/');

await testInstance.rebuild();

describe('Can print Eleventy data', { concurrent: true }, () => {
	const matrix = [
		['Static data', 'static'],
		['Computed data', 'computed'],
	];

	test.for(matrix)('%s', async ([_label, slug], { expect }) => {
		const result = testInstance.getBuildResultForUrl(`/data-handling/${slug}/`);
		await expect(result?.content).toMatchFileSnapshot(`./snapshots/data-${slug}.htmlsnap`);
	});
});

describe("Can include from Eleventy's `includes` directory", { concurrent: true }, () => {
	const matrix = [
		['Without data', 'without-data'],
		['With data', 'with-data'],
	];

	test.for(matrix)('%s', async ([_label, slug], { expect }) => {
		const result = testInstance.getBuildResultForUrl(`/include-handling/${slug}/`);
		await expect(result?.content).toMatchFileSnapshot(`./snapshots/includes-${slug}.htmlsnap`);
	});
});

describe("Can use layouts from Eleventy's `layouts` directory", { concurrent: true }, () => {
	const matrix = [
		['Example 1', 'example-1'],
		['Example 2', 'example-2'],
	];

	test.for(matrix)('%s', async ([_label, slug], { expect }) => {
		const result = testInstance.getBuildResultForUrl(`/layout-handling/${slug}/`);
		await expect(result?.content).toMatchFileSnapshot(`./snapshots/layouts-${slug}.htmlsnap`);
	});
});

describe('Can process Eleventy permalink keys', { concurrent: true }, () => {
	test('render a page at a nested permalink', async ({ expect }) => {
		const result = testInstance.getBuildResultForUrl('/page/at/nested/permalink/');
		await expect(result?.content).toMatchFileSnapshot('./snapshots/permalink-nested.htmlsnap');
	});

	test('render a page at root url', async ({ expect }) => {
		const result = testInstance.getBuildResultForUrl('/');
		await expect(result?.content).toMatchFileSnapshot('./snapshots/permalink-root.htmlsnap');
	});

	test('render a page at a dynamic permalink', async ({ expect }) => {
		const result = testInstance.getBuildResultForUrl('/directory-data-test-page.html');
		expect(result?.content).toBeDefined();
	});

	test('render a page at a dynamic permalink (js)', async ({ expect }) => {
		const result = testInstance.getBuildResultForUrl('/directory-data-test-page-js.html');
		expect(result?.content).toBeDefined();
	});

	test('skips rendering a file', async ({ expect }) => {
		const skipped = testInstance.getPageWithInput('\nLeave me out of this.\n');

		expect(skipped?.url).toBe(false);
	});
});

describe('Can run Vento filters', { concurrent: true }, () => {
	const matrix = [
		['Built-in methods', 'built-in'],
		['Chained methods', 'complex'],
		['Custom filter', 'custom'],
	];

	test.for(matrix)('%s', async ([_label, slug], { expect }) => {
		const result = testInstance.getBuildResultForUrl(`/filters/${slug}/`);
		await expect(result?.content).toMatchFileSnapshot(`./snapshots/filters-${slug}.htmlsnap`);
	});

	test('use context data (this.page)', async ({ expect }) => {
		const result = [
			testInstance.getBuildResultForUrl('/filters/with-this-example-1/'),
			testInstance.getBuildResultForUrl('/filters/with-this-example-2/'),
		].every((result) => result?.content === 'true');

		expect(result).toBe(true);
	});

	test('use context env (this.env)', async ({ expect }) => {
		const result = testInstance.getBuildResultForUrl('/filters/with-env-compile/');
		expect(result?.content).toBe('4');
	});
});

describe('Can run Eleventy shortcodes as Vento tags', { concurrent: true }, () => {
	const matrix = [
		['single shortcodes', 'single'],
		['single shortcodes with filters', 'single-filters'],
		['paired shortcodes', 'paired'],
		['paired shortcodes with filters', 'paired-filters'],
		['nested shortcodes with adjacent content', 'paired-nested'],
	];

	test.for(matrix)('run %s', async ([_label, slug], { expect }) => {
		const result = testInstance.getBuildResultForUrl(`/shortcodes/${slug}/`);
		await expect(result?.content).toMatchFileSnapshot(`./snapshots/shortcodes-${slug}.htmlsnap`);
	});
});

test('Can access collection data within templates', { concurrent: true }, async ({ expect }) => {
	const result = testInstance.getBuildResultForUrl(`/collections/iterate/`);
	await expect(result?.content).toMatchFileSnapshot(`./snapshots/collections-iterate.htmlsnap`);
});
