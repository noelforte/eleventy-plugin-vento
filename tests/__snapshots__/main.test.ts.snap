// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Can include from Eleventy's \`includes\` directory > With data > includes-with-data 1`] = `
"
<h1>Page with includes (data passing)</h1>

<section class="include">
	<p>This content is included.</p>
	<p>Passed data for \`extra\` key: 'super fancy data'</p>
</section>

"
`;

exports[`Can include from Eleventy's \`includes\` directory > Without data > includes-without-data 1`] = `
"
<h1>Page with includes</h1>

<section class="include">
	<p>This content is included.</p>
	<p>Passed data for \`extra\` key: ''</p>
</section>

"
`;

exports[`Can print Eleventy data > Computed data > data-computed 1`] = `
"
<h1>Dynamically generated from a custom key</h1>

<p>a custom key</p>
"
`;

exports[`Can print Eleventy data > Static data > data-static 1`] = `
"
Local Key:
<h1>Basic Page</h1>
<p>This is a standalone page.</p>

Directory Data:
<p>it's awesome!!</p>

Built-ins:
<p>Built with eleventy-plugin-vento</p>

Global Data:
<p>/data-handling/static/</p>

Global Data Function:
<p>Type of args[0]: 'bigint'<br>
Type of args[1]: 'boolean'<br>
Type of args[2]: 'function'<br>
Type of args[3]: 'number'<br>
Type of args[4]: 'object'<br>
Type of args[5]: 'string'<br>
Type of args[6]: 'symbol'<br>
Type of args[7]: 'undefined'</p>
"
`;

exports[`Can process Eleventy permalink keys > render a page at a nested permalink > permalink-nested 1`] = `
"<!doctype html>
<html lang="en">
	<head>
		<meta charset="UTF-8" />
		<meta name="viewport" content="width=device-width, initial-scale=1.0" />
		<title>Nested Page</title>
	</head>
	<body>
		<header>
			<h1>Page title: Nested Page</h1>
		</header>
		<main>
This content lives at \`/page/at/nested/permalink/\`
</main>
		<footer>
			<p>I am a footer, outside the layout</p>
		</footer>
	</body>
</html>
"
`;

exports[`Can process Eleventy permalink keys > render a page at root url > permalink-root 1`] = `
"<!doctype html>
<html lang="en">
	<head>
		<meta charset="UTF-8" />
		<meta name="viewport" content="width=device-width, initial-scale=1.0" />
		<title>Root page</title>
	</head>
	<body>
		<header>
			<h1>Page title: Root page</h1>
		</header>
		<main>
This should render on the homepage.
</main>
		<footer>
			<p>I am a footer, outside the layout</p>
		</footer>
	</body>
</html>
"
`;

exports[`Can run Eleventy shortcodes as Vento tags > run nested shortcodes with adjacent content > shortcodes-paired-nested 1`] = `
"<section class="some-class">

	Plaintext content

	Release 5 million possums!!!
	Release 15 million possums!!!
	Release 50 million possums!!!

	<pre class="code-block"><code class="lang-js">
		const text = 'This is some js code.';
	</code></pre>

	<section class="another-class">
		Some content
	</section>

	More plaintext content.

	<section>
		A third section, no class here though.
		<section>
			Sub-sub-section
		</section>
	</section>

	<section>
		ALL THE CAPITAL LETTERS!!
	</section>

</section>
"
`;

exports[`Can run Eleventy shortcodes as Vento tags > run paired shortcodes > shortcodes-paired 1`] = `
"Paired shortcode call:
<pre class="code-block"><code class="lang-text">
	This is some plaintext.
</code></pre>

Paired shortcode call with argument:
<pre class="code-block"><code class="lang-html">
	<h1>Who puts tags in all uppercase anyway?</h1>
</code></pre>

Paired shortcode call with page data:

	<p>This is a paragraph</p>
	<p>And this is another one</p>
	<p>One more for luck</p>

<span class="url">/shortcodes/paired/</span>

Paired shortcode call with explicit page data:
<section>
	<p>The page url is: /shortcodes/paired/</p>
	<p>The project description is: Adds support for the Vento templating language to Eleventy.</p>
	<p>The Eleventy Version is: 3.0.0</p>
	<p>The generator info is: Eleventy v3.0.0</p>
</section>
"
`;

exports[`Can run Eleventy shortcodes as Vento tags > run paired shortcodes with filters > shortcodes-paired-filters 1`] = `
"Paired shortcode call with filter:
<pre class="code-block"><code class="lang-text">
	This is some plaintext.
</code></pre>

Paired shortcode call with argument and filter:
<pre class="code-block"><code class="lang-html">
	<H1>WHO PUTS TAGS IN ALL UPPERCASE ANYWAY?</H1>
</code></pre>

Paired shortcode call with page data and filter:

	<P>THIS IS A PARAGRAPH</P>
	<P>AND THIS IS ANOTHER ONE</P>
	<P>ONE MORE FOR LUCK</P>

<span class="url">/shortcodes/paired-filters/</span>
"
`;

exports[`Can run Eleventy shortcodes as Vento tags > run single shortcodes > shortcodes-single 1`] = `
"Shortcode call:
Release the possums!!!

Shortcode call with argument:
Release 5 million possums!!!
Release 100 possums!!!

Shortcode call with different data types:
true

Shortcode call with page data:
page matched
"
`;

exports[`Can run Eleventy shortcodes as Vento tags > run single shortcodes with filters > shortcodes-single-filters 1`] = `
"Shortcode call with filter:
RELEASE THE POSSUMS!!!

Shortcode call with argument and filter:
RELEASE 5 MILLION POSSUMS!!!
RELEASE 100 POSSUMS!!!

Shortcode call with different data types and filter:
TRUE

Shortcode call with page data and filter:
PAGE MATCHED

"
`;

exports[`Can run Vento filters > Built-in methods > filters-built-in 1`] = `
"
Page title is 'A BASIC PAGE'

Custom props variable expressed as JSON:
{
  "colors": [
    "red",
    "blue",
    "green",
    "orange",
    "yellow"
  ],
  "awesome": true,
  "generator": {
    "name": "Eleventy",
    "version": "3.0.0"
  }
}
"
`;

exports[`Can run Vento filters > Chained methods > filters-complex 1`] = `
"

.god yzal eht revo spmuj xof nworb kciuq ehT
"
`;

exports[`Can run Vento filters > Custom filter > filters-custom 1`] = `
"Filter with defaults:
This is a <span>really cool</span> filter.

Filter with arguments:
This is a <strong>really cool</strong> filter.
"
`;

exports[`Can use layouts from Eleventy's \`layouts\` directory > Example 1 > layouts-example-1 1`] = `
"<!doctype html>
<html lang="en">
	<head>
		<meta charset="UTF-8" />
		<meta name="viewport" content="width=device-width, initial-scale=1.0" />
		<title>Page with layout</title>
	</head>
	<body>
		<header>
			<h1>Page title: Page with layout</h1>
		</header>
		<main>
<p>This is a page that uses the first layout.</p>

<p>This is another paragraph.</p>
<p>This is another paragraph.</p>
<p>This is another paragraph.</p>

</main>
		<footer>
			<p>I am a footer, outside the layout</p>
		</footer>
	</body>
</html>
"
`;

exports[`Can use layouts from Eleventy's \`layouts\` directory > Example 2 > layouts-example-2 1`] = `
"<!doctype html>
<html lang="en">
	<head>
		<meta charset="UTF-8" />
		<meta name="viewport" content="width=device-width, initial-scale=1.0" />
		<title>Page with layout</title>
	</head>
	<body>
		<header>
			<h1>Page title: Page with layout</h1>
		</header>
		<main>
			<section>This is an alternative layout.</section>
			<section>
<p>This is a page that uses the second layout.</p>

<p>This is another paragraph.</p>
<p>This is another paragraph.</p>
<p>This is another paragraph.</p>

</section>
		</main>
		<footer>
			<p>I am a footer, outside the layout</p>
		</footer>
	</body>
</html>
"
`;
